### PR TITLE
Metrics writer

### PR DESCRIPTION
**Pull request by sanych-sun**
PR to close the #23

Please pay attention to the breaking changes:
1) API changes in the ICollector and ICollectorRegistry
2) IOnDemandCollector was removed
3) changed the way to register default collectors
4) added dependency on NSubstitute 
5) removed unused anymore classes from Contracts (CSummary, CQuantile and CLabelPair will be reviewed/removed in upcoming pull requests)
6) MetricType enum was moved into root namespace

### CODE REVIEW

File: src/Prometheus.Client/Summary.cs [hunk 1]
```diff
@@ -174,17 +175,11 @@ internal void Init(ICollector parent, LabelValues labelValues, bool includeTimes
...
-            }
-
-            internal void Populate(CMetric cMetric, DateTime now)
-            {
-                cMetric.CSummary = new CSummary
-                {
-                    Quantiles = new CQuantile[_objectives.Count]
-                };
+                double count;
+                double sum;
```
Comment (phnx47) [inline src/Prometheus.Client/Summary.cs]: count & sum unused. it is bug?

**Comment by phnx47** [comment]
Not all contacts deleted. Any plans for this?

**Comment by phnx47** [comment]
Need come examples here: https://github.com/PrometheusClientNet/Prometheus.Client.Examples

File: src/Prometheus.Client/Collectors/ProcessStats/ProcessCollector.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            writer.WriteMetricHeader("process_num_threads", MetricType.Gauge, "Total number of threads");
+            writer.WriteSample(_process.Threads.Count);
+
+            writer.WriteMetricHeader("process_processid", MetricType.Untyped, "Process ID");
+            writer.WriteSample(_process.Id);
+
+            writer.WriteMetricHeader("process_start_time_seconds", MetricType.Untyped, "Start time of the process since unix epoch in seconds");
+            writer.WriteSample(_processStartTime);
+        }
+    }
```
Comment (phnx47) [inline src/Prometheus.Client/Collectors/ProcessStats/ProcessCollector.cs]: You use better MetricType, than previous release. Thank you!

**Comment by sanych-sun** [comment]
Yep,

I going to review summary-related classes and move them somewhere.
LabelValues - I believe could be replaced by KayValuePair<>.

**Pull request merged**